### PR TITLE
Fix WM_GETDPISCALEDSIZE message handling.

### PR DESCRIPTION
**Pull request by dreddy-work**
WM_GETDPISCALEDSIZE is sent once per device and when moved back to same device, we only get DPI_CHANGED message but not WM_GETDPISCALEDSIZE. Windows cache the result from the first time and reuse it. This isn't very clear from the [documentation](https://learn.microsoft.com/en-us/windows/win32/hidpi/wm-getdpiscaledsize) and is found while experimenting and validating the scenarios.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8018)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 1]
```diff
@@ -2191,28 +2191,23 @@ public virtual Font Font
...
 
-        private protected bool TryGetDpiFont(int dpi, [NotNullWhen(true)] out Font? font)
-        {
-            font = null;
-            if (!PInvoke.AreDpiAwarenessContextsEqualInternal(DpiAwarenessContext, DPI_AWARENESS_CONTEXT.DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2))
-            {
-                Debug.Assert(false, $"Fonts need to be cached only for PermonitorV2 mode applications : {DpiHelper.IsPerMonitorV2Awareness} : {DpiAwarenessContext}");
-                return false;
-            }
+            _dpiFonts ??= new Dictionary<int, Font>();
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:2207]: As you always create this, you might as well move it before the check for null above.

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
One minor comment

**Review by JeremyKuhne — DISMISSED** [review]

**Review by lonitra — APPROVED** [review]
LGTM

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 2]
```diff
@@ -2191,28 +2191,23 @@ public virtual Font Font
             }
         }
 
-        private protected void AddToDpiFonts(int dpi, Font font)
+        internal Font GetScaledFont(Font font, int newDpi, int oldDpi)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:2194]: Is it possible to use Control.DeviceDpi instead of oldDpi?

**Review by Tanya-Solyanik — COMMENTED** [review]
I'm not sure if Control.DeviceDpi can be used instead of the old DPI from the message

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs (hunk 2 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:2194]: in some cases, DeviceDPi is updated to the new DPI before calling this method. This new DPI is part of the message we are handling and should be more reliable if we ever reorder/restructure the code.

**Review by Tanya-Solyanik — APPROVED** [review]

**Comment by dreddy-work** [comment]
Thank you.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 3]
```diff
@@ -2191,28 +2191,23 @@ public virtual Font Font
...
-
-        private protected bool TryGetDpiFont(int dpi, [NotNullWhen(true)] out Font? font)
-        {
-            font = null;
-            if (!PInvoke.AreDpiAwarenessContextsEqualInternal(DpiAwarenessContext, DPI_AWARENESS_CONTEXT.DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2))
+            if (_dpiFonts.TryGetValue(newDpi, out Font? scaledFont))
             {
-                Debug.Assert(false, $"Fonts need to be cached only for PermonitorV2 mode applications : {DpiHelper.IsPerMonitorV2Awareness} : {DpiAwarenessContext}");
-                return false;
+                return scaledFont!;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:2202]: `!` shouldn't be required here

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:2202]: Thanks. Check that in next commit.

**Pull request merged**